src/framebuffer/swtfb_client.rs

```diff
@@ -0,0 +1,224 @@
+
+impl Default for SwtfbClient {
+    fn default() -> Self {
+        let msqid = unsafe {
+            libc::msgget(
+                SWTFB_MESSAGE_QUEUE_ID,
+                libc::IPC_CREAT | libc::SHM_R | libc::SHM_W,
+            )
+        };
+        assert!(msqid >= 0);
+
+        // Does nested need special handling?
+        // This may not be needed at all.
+        if env::var("RM2FB_ACTIVE").is_ok() {
+            env::set_var("RM2FB_NESTED", "1");
+        } else {
+            env::set_var("RM2FB_ACTIVE", "1");
+        }
+
+        Self {
+            msqid,
+            do_wait_ioctl: env::var("RM2FB_NO_WAIT_IOCTL").is_err(),
+        }
+    }
+}
+
+impl SwtfbClient {
+    pub fn open_buffer(&self) -> Result<(File, MmapRaw), IoError> {
+        let device = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/shm/swtfb.01")?;
+        let ret = unsafe { libc::ftruncate(device.as_raw_fd(), BUF_SIZE as libc::off_t) };
+        if ret < 0 {
+            return Err(IoError::last_os_error());
+        }
+        let mem_map = MmapOptions::new().len(BUF_SIZE as usize).map_raw(&device)?;
+        Ok((device, mem_map))
+    }
+
+    pub fn send(&self, update: &swtfb_update) -> bool {
+        unsafe {
+            let ptr = ptr::addr_of!(*update) as *const c_void;
+            libc::msgsnd(self.msqid, ptr, mem::size_of::<swtfb_update>(), 0) == 0
+        }
+    }
+
+    pub fn send_mxcfb_update(&self, update: &mxcfb_update_data) -> bool {
+        self.send(&swtfb_update {
+            mtype: MSG_TYPE::UPDATE_t,
+            data: swtfb_update_data { update: *update },
+        })
+    }
+
+    pub fn send_xochitl_update(&self, data: &xochitl_data) -> bool {
+        self.send(&swtfb_update {
+            mtype: MSG_TYPE::XO_t,
+            data: swtfb_update_data {
+                xochitl_update: *data,
+            },
+        })
+    }
+
+    pub fn wait_for_update_complete(&self) {
+        if !self.do_wait_ioctl {
+            return;
+        }
+
+        // https://github.com/ddvk/remarkable2-framebuffer/blob/1e288aa9/src/client/main.cpp#L149
+
+        let sem_name_str = format!("/rm2fb.wait.{}", unsafe { libc::getpid() });
+        let mut sem_name = [0u8; 512];
+        for (i, byte) in sem_name_str.as_bytes().iter().enumerate() {
+            sem_name[i] = *byte;
+        }
+        self.send_wait_update(&wait_sem_data { sem_name });
+        let sem_name_c = CString::new(sem_name_str.as_str()).unwrap();
+        let sem = unsafe { libc::sem_open(sem_name_c.as_ptr() as *const u8, libc::O_CREAT) };
+
+        let mut timeout = libc::timespec {
+            tv_nsec: 0,
+            tv_sec: 0,
+        };
+        unsafe {
+            libc::clock_gettime(libc::CLOCK_REALTIME, &mut timeout);
+        }
+        timeout.tv_nsec += SEM_WAIT_TIMEOUT;
+
+        if timeout.tv_nsec >= 1_000_000_000 {
+            timeout.tv_nsec -= 1_000_000_000;
+            timeout.tv_sec += 1;
+        }
```

Comment:
Shouldn't this be in a loop?
I have no idea if the clock here can jump around forward/backwards in large/small leaps...

Reply:
I hadn't bothered redoing this with rust types yet. Thought of this, too but rm2fb [did the same](https://github.com/ddvk/remarkable2-framebuffer/blob/1e288aa9/src/client/main.cpp#L175) and got away for it. For the sake of not needing to test this myself, the code is basically a copy for now.

I also had done some things like that for work. There are weird cases where such stuff is meant for leap seconds and rust's chrono and/or time crates handle conversion from timespec a lot more strictly (even though this just seems like a tame overflow counter). So didn't really wanna tamper with the already working solution. But your point seems pretty important as I know think about it more.

@raisjn any insights here? May be an issue for your client as well. Maybe this never mattered since the timeout never exceeded it in testing?

Reply:
For now I think we should just keep it as close to the original client as possible. If the client changes this, we can do so as well later. But it may be a bug that just got never triggered.

Reply:
I don't remember the reasoning here or why we did this, but a loop sounds right. there might be something in the git log for rm2fb about it (use blame?)

Reply:
The performance could probably be even faster/branchless when using something like this instead:

```diff
timeout.tv_nsec += SEM_WAIT_TIMEOUT;
+ timeout.tv_sec += timeout.tv_nsec / 1e9;
+ timeout.tv_nsec %= 1e9;
- if (timeout.tv_nsec >= 1e9) {
-   timeout.tv_nsec -= 1e9;
-   timeout.tv_sec++;
- }
```

Reply:
Added the above change with commit 3bd976cf.

To get this done here I opened a PR here for the original client to address it separately there: https://github.com/ddvk/remarkable2-framebuffer/pull/81